File: apple/NwcMobileApple/Sources/NwcMobileApple/NwcWakeExecution.swift
import Foundation

/// A non-sensitive NIP-47 method category used only for static presentation.
public enum NwcWakeRequestAction: Sendable, Equatable {
  case getInfo
  case getBalance
  case payInvoice
  case makeInvoice
  case lookupInvoice
  case listTransactions
}

/// Generic presentation guidance returned after Rust-owned wake policy runs.
///
/// The enum deliberately carries no invoice, amount, counterparty, relay, or
/// remote error text.
public enum NwcWakePresentationHint: Sendable, Equatable {
  case processing
  case completed
  case request(NwcWakeRequestAction)
  case openApplication
}

/// Cancellation object created for one bounded NSE execution.
public protocol NwcWakeCancellation: AnyObject, Sendable {
  func cancel()
}

/// Thin adapter implemented by the wallet using generated `NwcMobile` types.
///
/// Implementations should validate the payload through `validateWakeEnvelope`,
/// invoke `MobileNwcEngine.executeWake`, and map only its notification hint to
/// `NwcWakePresentationHint`.
public protocol NwcWakeExecutor: Sendable {
  func execute(
    payload: NwcWakePayload,
    executionMilliseconds: UInt64,
    cancellation: any NwcWakeCancellation
  ) async -> NwcWakePresentationHint
}

/// Creates a new native/Rust cancellation bridge for every request.
public typealias NwcWakeCancellationFactory =
  @Sendable () -> any NwcWakeCancellation

File: android/nwc-mobile/src/main/kotlin/org/nwc/mobile/android/NwcWakeExecution.kt
package org.nwc.mobile.android

/** Generic worker result with no invoice, amount, counterparty, or error text. */
enum class NwcWakeWorkerDisposition {
  COMPLETED,
  RETRY,
  OPEN_APPLICATION,
  REJECTED,
}

/** Request-scoped bridge to the generated Rust `MobileCancellation`. */
fun interface NwcWakeCancellation {
  fun cancel()
}

/**
 * Wallet-owned bridge to the generated `NwcMobile` Kotlin API.
 *
 * Implementations construct and validate `MobileWakeEnvelope`, invoke
 * `MobileNwcEngine.executeWake`, and map only the stable lifecycle result.
 */
fun interface NwcWakeExecutor {
  suspend fun execute(
    payload: NwcWakePayload,
    executionMilliseconds: Long,
    cancellation: NwcWakeCancellation,
  ): NwcWakeWorkerDisposition
}

internal enum class NwcPlatformWorkResult {
  SUCCESS,
  RETRY,
}

internal fun NwcWakeWorkerDisposition.platformResult(): NwcPlatformWorkResult =
  when (this) {
    NwcWakeWorkerDisposition.RETRY -> NwcPlatformWorkResult.RETRY
    NwcWakeWorkerDisposition.COMPLETED,
    NwcWakeWorkerDisposition.OPEN_APPLICATION,
    NwcWakeWorkerDisposition.REJECTED,
    -> NwcPlatformWorkResult.SUCCESS
  }
